dll/win32/shell32/CDefView.cpp

```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
 */
 BOOL CDefView::InitList()
 {
-    SHELLDETAILS sd;
-    WCHAR szTemp[50];
     HIMAGELIST big_icons, small_icons;
 
     TRACE("%p\n", this);
 
     m_ListView.DeleteAllItems();
 
+    Shell_GetImageLists(&big_icons, &small_icons);
+    m_ListView.SetImageList(big_icons, LVSIL_NORMAL);
+    m_ListView.SetImageList(small_icons, LVSIL_SMALL);
+
     m_hMenuArrangeModes = CreateMenu();
+    LoadColumns();
+    return TRUE;
+}
+
+/**********************************************************
+* Column handling
+*/
+static HRESULT SHGetLVColumnSubItem(HWND List, UINT Col)
+{
+    LVCOLUMN lvc;
+    lvc.mask = LVCF_SUBITEM;
+    if (!ListView_GetColumn(List, Col, &lvc))
+        return E_FAIL;
+    else
+        return lvc.iSubItem;
+}
+
+HRESULT CDefView::MapFolderColumnToListColumn(UINT FoldCol)
+{
+    // This function is only called during column management, performance is not critical.
+    for (UINT i = 0;; ++i)
+    {
+        HRESULT r = SHGetLVColumnSubItem(m_ListView.m_hWnd, i);
+        if ((UINT) r == FoldCol)
+            return i;
+        else if (FAILED(r))
+            return r;
+    }
+}
+
+HRESULT CDefView::MapListColumnToFolderColumn(UINT ListCol)
+{
+    // This function is called every time a LVITEM::iSubItem is mapped to
+    // a folder column (calls to GetDetailsOf etc.) and should be fast.
+    if (m_ListToFolderColMap)
+    {
+        UINT count = DPA_GetPtrCount(m_ListToFolderColMap);
+        if (ListCol < count)
+        {
+            HRESULT col = (INT)(INT_PTR) DPA_FastGetPtr(m_ListToFolderColMap, ListCol);
+            assert(col >= 0 && col == SHGetLVColumnSubItem(m_ListView.m_hWnd, ListCol));
```

Comment:
Better keep `col` to an `INT` or whatever. And have `CDefView::MapListColumnToFolderColumn` also return `INT`.